test/integration/http2_integration_test.cc

```diff
@@ -195,6 +195,104 @@ TEST_P(Http2IntegrationTest, Trailers) { testTrailers(1024, 2048); }
+                                                                  {":path", "/test/long/url"},
+                                                                  {":scheme", "http"},
+                                                                  {":authority", "host"}},
+                                          *response1);
+
+  fake_upstream_connection1 = fake_upstreams_[0]->waitForHttpConnection(*dispatcher_);
+  upstream_request1 = fake_upstream_connection1->waitForNewStream(*dispatcher_);
+
+  // Start request 2
+  response2.reset(new IntegrationStreamDecoder(*dispatcher_));
+  encoder2 = &codec_client_->startRequest(Http::TestHeaderMapImpl{{":method", "POST"},
+                                                                  {":path", "/test/long/url"},
+                                                                  {":scheme", "http"},
+                                                                  {":authority", "host"}},
+                                          *response2);
+  fake_upstream_connection2 = fake_upstreams_[0]->waitForHttpConnection(*dispatcher_);
+  upstream_request2 = fake_upstream_connection2->waitForNewStream(*dispatcher_);
+
+  // Finish request 1
+  codec_client_->sendData(*encoder1, request1_bytes, true);
+  upstream_request1->waitForEndStream(*dispatcher_);
+
+  // Finish request 2
+  codec_client_->sendData(*encoder2, request2_bytes, true);
+  upstream_request2->waitForEndStream(*dispatcher_);
+
+  // Respond to request 2
+  upstream_request2->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+  upstream_request2->encodeData(request2_bytes, true);
+  response2->waitForEndStream();
+  EXPECT_TRUE(upstream_request2->complete());
+  EXPECT_EQ(request2_bytes, upstream_request2->bodyLength());
+  EXPECT_TRUE(response2->complete());
+  EXPECT_STREQ("200", response2->headers().Status()->value().c_str());
+  EXPECT_EQ(request2_bytes, response2->body().size());
+
+  // Validate that idle time is not kicked in.
+  EXPECT_EQ(0, test_server_->counter("cluster.cluster_0.upstream_cx_idle_timeout")->value());
+  EXPECT_NE(0, test_server_->counter("cluster.cluster_0.upstream_cx_total")->value());
+
+  // Respond to request 1
+  upstream_request1->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+  upstream_request1->encodeData(request2_bytes, true);
+  response1->waitForEndStream();
+  EXPECT_TRUE(upstream_request1->complete());
+  EXPECT_EQ(request1_bytes, upstream_request1->bodyLength());
+  EXPECT_TRUE(response1->complete());
+  EXPECT_STREQ("200", response1->headers().Status()->value().c_str());
+  EXPECT_EQ(request2_bytes, response1->body().size());
+
+  // Do not send any requests and validate idle timeout kicks in after both the requests are done.
+  test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_idle_timeout", 1);
+
+  // Cleanup both downstream and upstream
+  codec_client_->close();
+  fake_upstream_connection1->close();
```

Comment:
Are the close calls needed here? Can we just move this up above line 286 and wait for disconnect due to idle timeout?

Reply:
This one still needs action?